### ISSUE TITLE
Allowing type specification through a map rather than a variadic string argument would be more flexible

### ISSUE THREAD

**Issue by tobgu** [kniren/gota#18]
Right now I have to either specify _all_ types or no types at all. Specifying the types in a `map[string]string` (`column name -> type name`) would add the possibility to specify types only for the columns you want to and fallback to auto typing for the other columns.

It could possibly also shorten the code in `ReadRecords` by simply checking if the column name is in the map and if not fallback to `findType`.

What do you think?

**Comment by kniren** [comment]
I think this is a good idea, however some consideration is needed. If we limit the argument to a `map[string]string` it is a better solution than now for an usability perspective but what happens if we want to force parsing all columns to a given type? What if we want to specify what does it mean by `NA` or other options that the `Read*` family of methods could need?

I'm all for simplicity and I found your solution to be better than the current implementation, but can we do better before we settle? Even since I read https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis I thought that this could be the way to go for such issues, but it is paramount that we are very mindful of the API that will go to the stable release.

I'm interested in knowing your opinion on this!

**Comment by tobgu** [comment]
Thanks for the link, I had not read it before. Feels like the technique has the potential to fill the void of not having default/named arguments for some use cases.

I don't think it would be a good idea to make the configuration functions operate on the dataframe or series themselves in this case since that would clutter those data structures with data that may only be of use initially, during construction.

Perhaps it would be good to have the functions operate on a configuration struct instead? So, for this particular case something along the lines of:

```
type Config struct {
     columnTypes map[string]string
}

func ColumnTypes(columnTypes map[string]string) func (*Config) error {
    return func ColumnTypes(c *Config) error {
         c.columnTypes = columnTypes
         return nil
    }
}

func ReadRecords(records [][]string, configs ...func (*Config) error) DataFrame {
    // New empty Config
    // Loop over configs applying each in turn, retur error if one of them fails
    ...
```

`Config` could be extended to hold more data and additional functions could be added to operate on it in the future if needed.

BTW: How come you return the error from the constructors as a field inside the DataFrame instead of an own value as usually seen in Go programs (eg. `func foo() (value, error)`)?

**Comment by kniren** [comment]
This feels like the best of both worlds solution. Good thinking!

The rationale behind returning the errors as a field inside DataFrame and
Series objects is that this enables pipes of operations as continuous
method chains `df.Filter().Filter(). Select()` so if an error happens at
any stage of the pipeline, the rest of the operations become a NoOp and the
error cascades until the end of the pipe. I feel this makes for a much
better API.

**Comment by tobgu** [comment]
Regarding the return value: I see, makes sense!

**Comment by kniren** [comment]
This has been implemented on the `dev` branch as of commit c6251c5741807e190d8f31d3d84404e384833fff

The LoadOptions type contains the following fields:

```
type LoadOptions struct {
	detectTypes bool
	hasHeader   bool
	types       map[string]Type
	defaultType Type
}
```

Subsequently, the functions to configure these options are:

- `CfgDetectTypes`
- `CfgHasHeader`
- `CfgColumnTypes `
- `CfgDefaultType `

As an example:

```
_ = LoadRecords(
	[][]string{
		{"A", "B", "C", "D"},
		{"a", "1", "true", "0"},
		{"b", "2", "true", "0.5"},
	},
	CfgHasHeader(true),
	CfgDetectTypes(true),
	CfgColumnTypes(map[string]Type{
		"A": String,
		"B": Int,
	}),
	CfgDefaultType(String),
)
```

But we can also call it with the default options:

```
_ = LoadRecords([][]string{...})
```

**Comment by tobgu** [comment]
Cool, hope to find some time to try it out soon.
